### PR TITLE
refactor: deactivate old quests on update

### LINKED ISSUE #66: [PoC] Refactor delete quest -> deactivate quest

### PR DESCRIPTION
**Pull request by guidota**
# Description

These changes propose deactivating quests instead of deleting them. In order to support that:
- on delete =>
   - deactivate: insert row in new table (deactivated_quests)
- on update =>
   - create a new quest
   - deactivate old quest
   - add an update (from -> to) in a new table (quest_updates)
   - quest instances already started will continue living with previous quest definitions
- on start  =>
   - only active quests can be started

Closes #66

### CODE REVIEW

File: crates/db/migrations/20230113195325_quests.sql [hunk 1]
```diff
@@ -1,28 +1,38 @@
...
 CREATE TABLE IF NOT EXISTS events ( 
   ID UUID PRIMARY KEY NOT NULL,
-  quest_instance_id UUID references quest_instances(ID) ON DELETE CASCADE, -- TODO: should not be a FK?
+  quest_instance_id UUID references quest_instances(ID),
   user_address TEXT NOT NULL,
   event bytea NOT NULL,
   timestamp TIMESTAMP NOT NULL DEFAULT now()
 );
 
+CREATE TABLE IF NOT EXISTS deactivated_quests (
```
Comment (lauti7) [inline crates/db/migrations/20230113195325_quests.sql:25]: Don't we already have a DB in Dev environment? If so, i think we should create another migration file

File: crates/db/migrations/20230113195325_quests.sql [hunk 2]
```diff
@@ -1,28 +1,38 @@
...
+  quest_id UUID references quests(ID), 
   user_address TEXT NOT NULL,
   start_timestamp TIMESTAMP NOT NULL DEFAULT now()
 );
 
--- Create events table
 CREATE TABLE IF NOT EXISTS events ( 
   ID UUID PRIMARY KEY NOT NULL,
-  quest_instance_id UUID references quest_instances(ID) ON DELETE CASCADE, -- TODO: should not be a FK?
+  quest_instance_id UUID references quest_instances(ID),
```
Comment (lauti7) [inline crates/db/migrations/20230113195325_quests.sql:19]: I know that from business logic (or application layer) there wouldn't be any deletion but I think we could keep `ON DELETE CASCADE`. Or do you prefer to remove it for preventing?

File: crates/db/migrations/20230113195325_quests.sql [hunk 3]
```diff
@@ -1,28 +1,38 @@
--- Create quests table
 CREATE TABLE IF NOT EXISTS quests (
   ID UUID PRIMARY KEY NOT NULL,
   name TEXT NOT NULL,
   description TEXT NOT NULL,
   definition bytea NOT NULL,
   created_at TIMESTAMP NOT NULL DEFAULT now(),
   updated_at TIMESTAMP NOT NULL DEFAULT now()
-  -- TODO: should we add the creator id (address)?
```
Comment (lauti7) [inline crates/db/migrations/20230113195325_quests.sql:9]: Shouldn't we add it to the table?

File: crates/db/src/lib.rs [hunk 4]
```diff
@@ -319,21 +326,59 @@ impl QuestsDatabase for Database {
     }
 }
 
+impl Database {
```
Comment (lauti7) [inline crates/db/src/lib.rs:329]: I don't understand why this new impl block

**Review by lauti7 — COMMENTED** [review]
Some comments

File: crates/db/migrations/20230113195325_quests.sql (hunk 1 above)
Comment (guidota) [reply crates/db/migrations/20230113195325_quests.sql:25]: Yes, but we can reinitialize it too

File: crates/db/migrations/20230113195325_quests.sql (hunk 3 above)
Comment (guidota) [reply crates/db/migrations/20230113195325_quests.sql:9]: Yes!

File: crates/db/src/lib.rs (hunk 4 above)
Comment (guidota) [reply crates/db/src/lib.rs:329]: There are some internal methods there to support transactions that can't be added to the other `impl` block.

File: crates/db/migrations/20230113195325_quests.sql (hunk 2 above)
Comment (guidota) [reply crates/db/migrations/20230113195325_quests.sql:19]: I would leave it without `ON DELETE CASCADE` for now as we should not be deleting things manually.

File: crates/db/src/lib.rs (hunk 4 above)
Comment (lauti7) [reply crates/db/src/lib.rs:329]: got it

**Review by lauti7 — APPROVED** [review]
LGTM 💯 ! just fix one Clippy warning

**Pull request merged**